src/member-api.js

```diff
@@ -54,42 +75,148 @@ export class MemberApi extends TypedEmitter {
+      roleDescription,
+      timeout = 1_800_000,
+    }
+  ) {
     assert(isRoleIdForNewInvite(roleId), 'Invalid role ID for new invite')
+    assert(
+      !this.#deviceIdsWithPendingInvites.has(deviceId),
+      'Already inviting this device ID'
+    )
+    assert(timeout > 0, 'Timeout should be a positive number')
+    assert(timeout < 2 ** 32, 'Timeout is too large')
 
-    const { name: deviceName } = await this.getById(this.#ownDeviceId)
+    /**
+     * These cleanup functions should be synchronous so we don't have to
+     * `await` calls to `handleAborted`.
+     * @type {Array<() => void>}
+     */
+    const cleanupFns = []
+    const cleanup = () => {
+      for (const fn of cleanupFns) {
+        try {
+          fn()
+        } catch (_err) {
+          console.error('Cleanup function failed')
+        }
+      }
+    }
 
-    // since we are always getting #ownDeviceId,
-    // this should never throw (see comment on getById), but it pleases ts
-    if (!deviceName)
-      throw new Error(
-        'Internal error trying to read own device name for this invite'
-      )
+    try {
+      let handleAborted = noop
+      const timeoutId = setTimeout(() => {
+        handleAborted = () => {
+          throw new Error('Invite timed out')
+        }
+      }, timeout)
+      cleanupFns.push(() => {
+        clearTimeout(timeoutId)
+      })
 
-    const projectId = projectKeyToId(this.#projectKey)
-    const project = await this.#dataTypes.project.getByDocId(projectId)
+      this.#deviceIdsWithPendingInvites.add(deviceId)
+      cleanupFns.push(() => {
+        this.#deviceIdsWithPendingInvites.delete(deviceId)
+      })
 
-    if (!project.name)
-      throw new Error('Project must have a name to invite people')
+      const { name: invitorName } = await this.getById(this.#ownDeviceId)
+      // since we are always getting #ownDeviceId,
+      // this should never throw (see comment on getById), but it pleases ts
+      if (!invitorName)
+        throw new Error(
+          'Internal error trying to read own device name for this invite'
+        )
+      handleAborted()
 
-    const response = await this.#rpc.invite(deviceId, {
-      projectKey: this.#projectKey,
-      encryptionKeys: this.#encryptionKeys,
-      projectInfo: { name: project.name },
-      roleName: roleName || ROLES[roleId].name,
-      roleDescription,
-      invitorName: deviceName,
-      timeout,
-    })
+      const projectId = projectKeyToId(this.#projectKey)
+      const projectPublicId = projectKeyToPublicId(this.#projectKey)
+      const project = await this.#dataTypes.project.getByDocId(projectId)
+      const projectName = project.name
+      if (!projectName)
+        throw new Error('Project must have a name to invite people')
+      handleAborted()
 
-    if (response === InviteResponse_Decision.ACCEPT) {
-      await this.#roles.assignRole(deviceId, roleId)
-    }
+      const inviteId = crypto.randomBytes(32)
+
+      const {
```

Comment:
Again I think there is a better pattern than deferred promises, but it's ok for now.

Reply:
Addressed in 0a625010c5117cf285103fd9d64951280405bc71.